src/fieldAnalyser/fieldAnalyser.ts

```diff
@@ -49,75 +99,58 @@ export class FieldAnalyser {
   }
 
-  private getMissingFieldsFromOrg(
-    batch: BatchUpdateDocuments,
-    alreadyCreatedFields: string[]
-  ): Record<string, FieldTypeDict> {
-    const missingFieldsFromOrg: Record<string, FieldTypeDict> = {};
-
-    batch.addOrUpdate.flatMap((doc: DocumentBuilder) => {
-      const documentMetadata = Object.entries({...doc.build().metadata});
-
-      for (const [metadataKey, metadataValue] of documentMetadata) {
-        if (alreadyCreatedFields.includes(metadataKey)) {
-          continue;
-        }
-        const metadataType = this.getValue(metadataValue);
-        if (missingFieldsFromOrg[metadataKey]) {
-          // Possible metadata inconsistency
-          let fieldType = missingFieldsFromOrg[metadataKey][metadataType];
-          missingFieldsFromOrg[metadataKey][metadataType] = fieldType
-            ? ++fieldType
-            : 1;
-        } else {
-          missingFieldsFromOrg[metadataKey] = {[metadataType]: 1};
-        }
-      }
-    });
-
-    return missingFieldsFromOrg;
+  private async ensureExistingFields(): Promise<string[]> {
+    if (this.existingFields === undefined) {
+      this.existingFields = await this.listAllFieldsFromOrg();
+    }
+    return this.existingFields;
   }
 
-  private getFieldTypes(
-    missingFieldsFromOrg: Record<string, FieldTypeDict>
-  ): Record<string, FieldTypes> {
-    const fieldsToCreate: Record<string, FieldTypes> = {};
+  private getFieldTypes(): FieldBuilder {
+    const fieldBuilder = new FieldBuilder();
 
-    Object.entries(missingFieldsFromOrg).map(([fieldName, fieldTypeDict]) => {
-      this.storePossibleTypeInconsistencies(fieldName, fieldTypeDict);
-      const fieldType = this.getMostProbableType(fieldTypeDict);
-      fieldsToCreate[fieldName] = fieldType;
+    this.missingFields.forEach((fieldTypeMap, fieldName) => {
+      this.storePossibleTypeInconsistencies(fieldName, fieldTypeMap);
+      const fieldType = this.getMostProbableType(fieldTypeMap);
+      fieldBuilder.add(fieldName, fieldType);
     });
 
-    return fieldsToCreate;
+    return fieldBuilder;
   }
 
   private storePossibleTypeInconsistencies(
     fieldName: string,
-    fieldTypeDict: FieldTypeDict
+    fieldTypeMap: FieldTypeMap
   ) {
-    const typePossibilities = Object.keys(fieldTypeDict) as FieldTypes[];
-    if (typePossibilities.length > 1) {
-      this.inconsistentFields.add(fieldName, typePossibilities);
+    if (fieldTypeMap.size > 1) {
+      const fieldTypes = Array.from(fieldTypeMap.keys());
+      this.inconsistencies.add(fieldName, fieldTypes);
     }
   }
 
-  private getMostProbableType(field: FieldTypeDict): FieldTypes {
-    const [fieldType] = Object.entries(field).reduce((previous, current) => {
-      if (current[1] > previous[1]) {
-        return current;
-      } else if (current[1] < previous[1]) {
-        return previous;
-      } else {
-        return [current, previous].sort(this.typeCompare)[0];
+  private getMostProbableType(field: FieldTypeMap): FieldTypes {
```

Comment:
Here's what I think is a simpler implementation of the method:

```suggestion
  private getMostProbableType(field: FieldTypeMap): FieldTypes {
    const sortedType = Array.from(field.entries()).sort((a, b) => {
      const countDiff = a[1] - b[1];
      return countDiff ? countDiff : this.getFieldTypes(a[0], b[0]);
    })

    return sortedType[0][0];
  }
```

Reply:
True. A little less performant than the reduce implementation but way more readable.
Nice

Reply:
I compared both implementations with 10000 fields to create.
The `reduce` algorithm takes ~8 seconds vs ~12 seconds (algorithm from your example). 
No big difference in relation to the total amount of time to push documents.